README.md
# Python DVB Companion Screen Synchronisation protocol library and tools

[![Build status](https://travis-ci.org/bbc/pydvbcss.svg?branch=0.5.2)](https://travis-ci.org/bbc/pydvbcss)
[![Docs Status
(0.5.2)](https://readthedocs.org/projects/pydvbcss/badge/?version=0.5.2)](http://pydvbcss.readthedocs.io/en/0.5.2/?badge=0.5.2)
[![Docs Status (latest)](https://readthedocs.org/projects/pydvbcss/badge/?version=latest)](http://pydvbcss.readthedocs.io/en/latest/?badge=latest)
[![Latest PyPI package](https://img.shields.io/pypi/v/pydvbcss.svg)](https://pypi.python.org/pypi/pydvbcss)

* **[How to install](#install-the-code)**
* **[Read the documentation](#read-the-documentation)**
* **[Run the examples](#run-the-examples)**

**pydvbcss** is a set of Python 2.7 libraries and command-line tools that implement some of the
protocols defined in the DVB CSS specification (published as [ETSI 103-286 part 2](http://www.etsi.org/standards-search?search=103+286&page=1&title=1&keywords=1&ed=1&sortby=1))
and are used for the "inter-device synchronisation" feature in **[HbbTV 2](http://hbbtv.org/resource-library/)**.
These protocols enable synchronisation of media presentation between a TV
and Companion devices (mobiles, tablets, etc).

This library includes simple to use high level abstractions that wrap up the
server or client behaviour for each protocol as well as low level code for
packing and unpacking messages sent across the protocols. There are also
objects that work with the rest of the library to represent clocks and timelines.

This code is intended as an informal reference and is suitable for building
prototypes and testing tools that implement TV (server) or Companion
(client) behaviour. It is not considered production ready or suitable for
integration into consumer products.

The code does not implement media playback functionality and this is not a planned
feature.

The DVB CSS specification was formerly published as [DVB Bluebook A167-2](https://www.dvb.org/search/results/keywords/A167). This is deprecated in favour of the [ETSI spec](http://www.etsi.org/standards-search?search=103+286&page=1&title=1&keywords=1&ed=1&sortby=1).

## Getting started

**pydvbcss** requires [ws4py](https://ws4py.readthedocs.io/en/latest/) for
use in clients and servers, and also [cherrypy](http://www.cherrypy.org)
for server implementations.  The steps below describe how to install these.

**pydvbcss** has been developed on Mac OS X 10.10 but has also been used
successfully on Microsoft Windows 7 and Ubuntu 14.04.



### Read the Documentation

The docs for the library can be read online on readthedocs.org:

   * [![Docs Status (0.5.2)](https://readthedocs.org/projects/pydvbcss/badge/?version=0.5.2)](http://pydvbcss.readthedocs.io/en/0.5.2/?badge=0.5.2) [Docs for 0.5.2 release](http://pydvbcss.readthedocs.io/en/0.5.2/?badge=0.5.2)
   
   * [![Docs Status (latest)](https://readthedocs.org/projects/pydvbcss/badge/?version=latest)](http://pydvbcss.readthedocs.io/en/latest/?badge=latest) [Docs for latest commits to master release](http://pydvbcss.readthedocs.io/en/latest/?badge=latest)

Links are also available from those pages through to documentation for earlier releases.



### Install the code ...

*On Mac OS X and Linux you may need to run one or more of the commands as root.*

#### Using PyPi _(core library only, no examples or tools)_

   * [![Latest PyPI package](https://img.shields.io/pypi/v/pydvbcss.svg)](https://pypi.python.org/pypi/pydvbcss) [See the pydvbcss PyPI package page](https://pypi.python.org/pypi/pydvbcss). 

If you ONLY want the library (not the [code examples and tools](#run-examples) ) and
if you don't require the very latest bugfixes, then you can install a recent
release package from the Python Package Index (PyPI) using
[pip](https://pip.pypa.io/en/latest/installing.html):

    $ pip install pydvbcss

Or if upgrading from a previous version:

    $ pip install --upgrade pydvbcss

You can use `pip search pydvbcss` to verify which version is installed.

> *See note in the next section about `CherryPy` and `ws4py` dependencies.*


#### From Github or a release tarball _(includes examples and tools)_

The [master branch](https://github.com/BBC/pydvbcss/tree/master) is the latest
state of the code, including any recent bugfixes. It is mostly stable but
might have occasional small API changes.
[Release snapshots](https://github.com/BBC/pydvbcss/releases) are also available
but won't contain the very latest bugfixes or new features.
Both of these options include the full code, including [examples](#run-examples).

First you need to install dependencies...

We recommend using [pip](https://pip.pypa.io/en/latest/installing.html) to install
dependencies from the Python Package Index [PyPI](https://pypi.python.org/pypi):

    $ pip install -r requirements.txt

> *NOTE: There have been recent incompatibilities between certain versions of
> `cherrypy`, `ws4py` and `cheroot`. Therefore, `requirements.txt` requires specific
> (older) versions of these pacakges. You are welcome to try newer versions installing
> them manually. See [#15](https://github.com/bbc/pydvbcss/issues/15) for background
> details.*

Then take (or update) your clone of the repository *master* branch, or
download and unzip a snapshot release and run the `setup.py` script to
install:

    $ python setup.py install
    
This will install all module packages under 'dvbcss'.

There is a limited test suite (it only tests certain classes at the moment).
Run it via setup.py:

    $ python setup.py test

This checks some timing sensitive implementation issues, so ensure you are not
running any CPU intensive tasks at the time.



## Running the examples and tools

There is a set of example and tools demonstrating simple servers and clients for the
protocols included with the library. See the 
[quick start guide](https://BBC.github.io/pydvbcss/docs/latest/examples.html) 
in the documentation to see how to run them.

The clients are useful tools to test a TV implementation is outputting the correect data.

The servers can be modified to simulate a TV that is playing content with an ID
and timeline(s) that a companion application expects.

### Example: checking protocols implemented by a TV

Start the content playing on the TV and ensure it is serving the protocols (for HbbTV 2
TVs this requires an HbbTV application to enable inter-device synchronisation).

Suppose the TV is serving the CII protocol at the URL `ws://192.168.0.57:7681/cii`...

To check the CII protocol:

    $ python examples/CIIClient.py ws://192.168.0.57:7681/cii

Suppose that the messages returned report the URL of the TS protocol endpoint as being `ws://192.168.0.57:7681/ts` and the wall clock protocol as being at `192.168.0.57` port `6677`...

To check the TV's Wall Clock protocol:`

    $ python examples/WallClockClient.py 192.168.0.57 6677

To check the TV reporting a PTS timeline (uses both Wall Clock and TS protocols):

    $ python examples/TSClient.py ws://192.168.0.57:7681/ts \
        udp://192.168.0.57:6677 \
        "" \
        "urn:dvb:css:timeline:pts" \
        9000


## Super-quick introduction to the protocols

DVB has defined 3 protocols for communicating between a companion and TV in
order to create synchronised second screen / dual screen / companion
experiences (choose whatever term you prefer!) that are implemented here:

* CSS-CII - A WebSockets+JSON protocol that conveys state from the TV, such
  as the ID of the content being shown at the time. It also carries the URLs
  to connect to the other two protocols.

* CSS-WC - A simple UDP protocol (like NTP but simplified) that establishes
  a common shared clock (a "wall clock") between the TV and companion,
  compensating for network delays.

* CSS-TS - Another WebSockets+JSON protocol that communicates timestamps
  from TV to Companion that describe the current timeline position.

The TV implements servers for all 3 protocols. The Companion implements
clients.

There are other protocols defined in the specification (CSS-TE and CSS-MRS) that
are not currently implemented by this library.


## Building the documentation for yourself

You can also build the documentation yourself. It is written using the
[sphinx](http://www.sphinx-doc.org) documentation build system.

Building the documentation requires [sphinx](http://www.sphinx-doc.org) and
the sphinx "read the docs" theme. The easiest way is using PyPI:

    $ pip install sphinx
    $ pip install sphinx_rtd_theme

The `docs` directory contains the configuration and main documentation
sources that descibe the structure. Most of the actual words are in the
inline docstrings in the source code. These structural pages pull these in.

To build docs in HTML format, either:

    $ python setup.py build_sphinx

or:

    $ cd docs
    $ make html
    


## Contact and discuss

Discuss and ask questions on the [pydvbcss google group](<https://groups.google.com/forum/#!forum/pydvbcss>).

The original author is Matt Hammond 'at' bbc.co.uk



## Licence

All code and documentation is licensed under the Apache License v2.0.



## Contributing

If you would like to contribute to this project, see
[CONTRIBUTING](CONTRIBUTING.md) for details.



CHANGELOG.md
# Change Log

## Latest

# 0.5.2 : pypi packaging bugfix

This is a minor release that fixes a packaging bug that may cause pydvbcss
to be unable to install correctly unless CherryPy and ws4py packages are
already installed.

* Fixed missing `requirements.txt` in pypi package
  ([c696395](https://github.com/bbc/pydvbcss/commit/c6963955f87e9569b6de689697ad95fe182e1893))


# 0.5.1 : Installation bugfixes

This is a minor release containing a temporary workaround for installation
problems caused by changes to packages this library depends on (cherrypy).

* Temporary fix for bug due to changes in cherrypy
  ([f5bb1b1](https://github.com/bbc/pydvbcss/commit/f5bb1b17653f4b431adadb0ad7f819cd4c444640) and
   [7ad71e9](https://github.com/bbc/pydvbcss/commit/7ad71e9c96f4732666725455b30767e9daa891cd)).
  Will be reverted once resolved by dependencies


## 0.5.0 : CII server improvement and fix for ws4py 0.3.6

This is a minor release containing mostly maintenance fixes arising from
recent changes to cherrypy/ws4py, some small documentation improvements,
and an enhancement to CII server support for when bound to multiple
network interfaces.

* Enhancement: CII server can transparently rewrite `wcUrl`, `tsUrl` and
  `teUrl` to take account of the network interface the client has connected
  to - enabling it to work when bound to multiple interfaces.
  ([#14](https://github.com/bbc/pydvbcss/issues/14))
* Docs: `README.md` improvements.
* Switch from temporary to permanent fix due to cherrypy/ws4py bugs that have now been
  fixed. ([#13](https://github.com/bbc/pydvbcss/issues/13))

## 0.4.1 : Bugfixes and new clock object type

This is a minor release that includes a temporary workaround for a problem
ith dependencies (cherrypy and ws4py) when installing. It also includes
various bug fixes and a new type of Clock object that is useful when
compensating for rendering latencies.

* Bugfix: fixed semantics on `calcWhen()`, `toParenTicks()` and `toRootTicks()`
  when clock speed is 0.` ([32256b6](https://github.com/bbc/pydvbcss/commit/32256b6f94ce01466ab4645097672e58a00456cc))
* API addition: new `OffsetClock` class to make it easy to compensate for rendering  latencies.
* Temporary fix for bug due to changes in cherrypy ([f8ec3f6](https://github.com/bbc/pydvbcss/commit/f8ec3f665896d8cc90c5963781e00d366c9d8331)). Will be reverted once resolved by dependencies.

## 0.4.0 : Bugfixes and overhaul of clock object model (new-clock-model)

This release contains a significant internal upgrade to the `dvbcss.clock`
module with minor knock-on effects on other packages - particularly wall clock
client and server code and algorithms. The changes have been implemented to
be mostly backwardly compatible, so existing code should continue to work. The
only exception will be any custom wall clock client algorithms.

Clock objects can now calculate **dispersion** (error bounds). Clocks can also
track clock **availability** (mirroring the concept of timeline availability).

Wall clock client algorithms have been switched from measuring and adjusting
the same clock (representing the wall clock) to instead measuring its parent
and setting the correlation. Any custom wall clock client algorithms will
not work and will need to be updated.

#### How to 'upgrade' existing code

* Use a `CorrelatedClock` instead of a `TunableClock` to model a wall clock.
  and set the "maximum frequency error" when initialising the `SysClock`
  instead of passing it to a `WallClockClient` or `WallClockServer`.
  
* Use a `Correlation` object instead of a tuple (parentT,childT) to represent
  correlations for a `CorrelatedClock`
  
* Control and check timeline availability of clock objects instead of setting
  or querying the availability of the TS protocol server or client.
  
* Update any custom algorithms you might have created for wall clock clients.

  * `Candidate` objects now represent the relationship between the **parent**
    of the local wall clock (instead of the local wall clock itself)
    and the server's wall clock.`
  
  * Algorithms and Filters now receive a single `Candidate` object in units of
    nanoseconds. They no longer receive a `Candidate` converted to units of
    clock ticks.
  
  * Predictors should return a `Correlation` instead of an adjustment value.
  
* `DispersionCalculator` is being deprecated. It still exists, but you should
  stop using it and instead use `Candidate.calcCorrelationFor` function to 
  create a correlation which you then put into a `CorrelatedClock` and call
  the `dispersionAtTime` function.
  
Examples of old way (0.3.x and earlier):
  
    s = SysClock()
    wallClock = TunableClock(s, tickRate=1000000000)

    algorithm = LowestDispersionCandidate(wallClock,repeatSecs=1,timeoutSecs=0.5, localMaxFreqErrorPpm=500)
    wc_client=WallClockClient(bind, dest, wallClock, algorithm)
    wc_client.start()

    timeline = CorrelatedClock(wallClock, tickRate=1000, (10,20))

    timeline.correlation = ( timeline.correlation[0], timeline.correlation[1] + 50)

    ts = TSClientClockController(tsUrl, contentIdStem, timelineSelector, timeline)
    print ts.available

Equivalent new way (0.4 and later):
  
    s=SysClock(maxFreqErrorPpm=500)
    wallClock=CorrelatedClock(s,tickRate=1000000000)

    algorithm = LowestDispersionCandidate(wallClock,repeatSecs=1,timeoutSecs=0.5)
    wc_client=WallClockClient(bind, dest, wallClock, algorithm)
    wc_client.start()

    timeline = CorrelatedClock(wallClock, tickRate=1000, Correlation(10,20))

    timeline.correlation = timeline.correlation.butWith(childTicks=timeline.childTicks + 50)

    ts = TSClientClockController(tsUrl, contentIdStem, timelineSelector, timeline)
    print timeline.isAvailable()

#### Summary of changes:

Main changes in new clock model ([da846a9](https://github.com/bbc/pydvbcss/commit/da846a96ec8dd3e23b4a3e363fd98d1c495cc8c5)):
* API addition: can setParent() on `CorrelatedClock`, `RangeCorrelatedClock` and `TunableClock`.

examples/TSClient.py
#!/usr/bin/env python
#
# Copyright 2015 British Broadcasting Corporation
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""\
.. py:module:: examples.TSClient

    It works by implementing `both` a wall clock client and a CSS-TS client. A
    :class:`~dvbcss.protocol.client.ts.TSClientClockController` object is instantiated
    and provided with a :class:`~dvbcss.clock.CorrelatedClock` object to represent
    the synchronisation timeline. The controller adjusts the clock object to match
    the timeline information coming from the server.

    At the command line you must specify:
    
    * the WebSocket URL of the CSS-TS server, in the form `ws://<host>:<port>/<path>`
    * a `udp://<host>:<port>` format URL for the Wall Clock server
    * The content ID stem and timeline selector to be used when requesting the timeline
    * The tick rate of the timeline.
    
    Default options can be overridden for the IP address and port that the Wall Clock client
    binds to and to reduce the amount of logging output.
    
    Use the ``--help`` command line option for usage information.
    
"""

if __name__ == "__main__":
    
    import _useDvbCssUninstalled  # Enable to run when dvbcss not yet installed ... @UnusedImport

    from dvbcss.clock import SysClock
    from dvbcss.clock import CorrelatedClock
    from dvbcss.clock import Correlation
    from dvbcss.protocol.client.wc import WallClockClient
    from dvbcss.protocol.client.wc.algorithm import LowestDispersionCandidate

    from dvbcss.protocol.client.ts import TSClientClockController
    
    import dvbcss.monotonic_time as time
    import logging
    import argparse
    import dvbcss.util
    import sys
    
    DEFAULT_WC_BIND=("0.0.0.0","random")
    
    parser=argparse.ArgumentParser(
        description="Run a DVB TM-CSS Wall Clock Client (WC-Client) and connection to CSS-TS to obtain a timeline.")

    parser.add_argument("-q","--quiet",dest="quiet",action="store_true",default=False,help="Suppress extraneous output during runtime. Overrides loglevel option")
    parser.add_argument("--loglevel",dest="loglevel",action="store",type=dvbcss.util.parse_logLevel, nargs=1, help="Set logging level to one of: critical, error, warning, info, debug. Default=info",default=[logging.INFO])
    parser.add_argument("--wcloglevel",dest="wcloglevel",action="store",type=dvbcss.util.parse_logLevel, nargs=1, help="Set logging level for the wall clock client to one of: critical, error, warning, info, debug. Default=warn",default=[logging.WARN])
    parser.add_argument("tsUrl", action="store", type=dvbcss.util.wsUrl_str, nargs=1, help="ws:// URL of CSS-TS end point")
    parser.add_argument("wcUrl", action="store", type=dvbcss.util.udpUrl_str, nargs=1, help="udp://<host>:<port> URL of CSS-WC end point")
    parser.add_argument("contentIdStem", action="store", type=str, nargs=1, help="contentIdStem")
    parser.add_argument("timelineSelector", action="store", type=str, nargs=1, help="Timeline selector")
    parser.add_argument("timelineFreq", action="store", type=int, nargs=1, help="Ticks per second of the media timeline")
    parser.add_argument("wc_bindaddr",action="store", type=dvbcss.util.iphost_str, nargs="?",help="IP address or host name to bind WC client to (default="+str(DEFAULT_WC_BIND[0])+")",default=DEFAULT_WC_BIND[0])
    parser.add_argument("wc_bindport",action="store", type=dvbcss.util.port_int_or_random,   nargs="?",help="Port number to bind WC client to (default="+str(DEFAULT_WC_BIND[1])+")",default=DEFAULT_WC_BIND[1])
    args = parser.parse_args()

    tsUrl = args.tsUrl[0]
    wc_dest=args.wcUrl[0]
    wc_bind=(args.wc_bindaddr, args.wc_bindport)
    contentIdStem = args.contentIdStem[0]
    timelineSelector= args.timelineSelector[0]
    timelineFreq = args.timelineFreq[0]
    
    if args.quiet:
        logging.disable(logging.CRITICAL)
    else:
        logging.basicConfig(level=args.loglevel[0])

    logging.getLogger("dvbcss.protocol.client.wc").setLevel(args.wcloglevel[0])

    sysclock=SysClock()
    wallClock=CorrelatedClock(sysclock,tickRate=1000000000) # nanos
    
    algorithm = LowestDispersionCandidate(wallClock,repeatSecs=1,timeoutSecs=0.5)
    
    wc_client=WallClockClient(wc_bind, wc_dest, wallClock, algorithm)
    wc_client.start()
    
    timelineClock = CorrelatedClock(wallClock, timelineFreq)
    timelineClock.setAvailability(False)

    print "Connecting, requesting timeline for:"
    print "   Any contentId beginning with:",contentIdStem
    print "   and using timeline selector: ",timelineSelector
    print

    ts = TSClientClockController(tsUrl, contentIdStem, timelineSelector, timelineClock, correlationChangeThresholdSecs=0.001)
    
    exiting=False
    tsClientLogger = logging.getLogger("TSClient")
    def reportCallback(msg,exit=False):
        def callback(*a,**k):
            global exiting
            tsClientLogger.info(msg+"\n")
            if exit:
                exiting=True
                wc_client.stop()
                sys.exit(0)
        return callback
    
    ts.onConnected           = reportCallback("connected")
    ts.onDisconnected        = reportCallback("disconnected",exit=True)
    ts.onTimelineAvailable   = reportCallback("timeline became available")
    ts.onTimelineUnavailable = reportCallback("timeline became un-available")
    ts.onTimingChange        = reportCallback("change in timing and/or play speed")
    
    ts.connect()
    while not exiting:
        time.sleep(0.4)
        print ts.getStatusSummary(),
        print "   Uncertainty (dispersion) = +/- %0.3f milliseconds" % (algorithm.getCurrentDispersion()/1000000.0)
